File: wisefy/src/androidTest/java/com/isupatches/wisefy/main/GetIIPTests.java
package com.isupatches.wisefy.main;

import static com.isupatches.wisefy.TestUtils.TEST_IP_ADDRESS_STRING;
import static com.isupatches.wisefy.TestUtils.VERIFICATION_SUCCESS_TIMEOUT;

import static org.junit.Assert.assertEquals;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;

import com.isupatches.wisefy.AbstractBaseAndroidJUnit4TestClass;
import com.isupatches.wisefy.WiseFy;
import com.isupatches.wisefy.callbacks.GetIPCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodeDefs;

import org.junit.Test;

/**
 * Tests the ability to get the IP of a device.
 *
 * @see WiseFy#getIP()
 * @see WiseFy#getIP(GetIPCallbacks)
 *
 * @author Patches
 */
public class GetIIPTests extends AbstractBaseAndroidJUnit4TestClass {

  public GetIIPTests() {
    // No-op
  }

  @Test
  public void sync_getIP_failure_missingPrerequisites() {
    getMockWiseFyPrechecksUtil().geIP_failure();
    assertEquals(null, getWiseFy().getIP());
  }

  @Test
  public void sync_getIP_failure() {
    getMockNetworkUtil().ip_failure();
    assertEquals(null, getWiseFy().getIP());
  }

  @Test
  public void sync_getIP_success() {
    getMockNetworkUtil().ip_success();
    assertEquals(TEST_IP_ADDRESS_STRING, getWiseFy().getIP());
  }

  @Test
  public void async_getIP_failure_missingPrerequisites() {
    getMockWiseFyPrechecksUtil().geIP_failure();
    final GetIPCallbacks mockCallbacks = mock(GetIPCallbacks.class);
    getWiseFy().getIP(mockCallbacks);
    verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getIPWiseFyFailure(WiseFyCodeDefs.MISSING_PREREQUISITE);
  }

  @Test
  public void async_getIP_failure() {
    getMockNetworkUtil().ip_failure();
    final GetIPCallbacks mockCallbacks = mock(GetIPCallbacks.class);
    getWiseFy().getIP(mockCallbacks);
    verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureRetrievingIP();
  }

  @Test
  public void async_getIP_success() {
    getMockNetworkUtil().ip_success();
    final GetIPCallbacks mockCallbacks = mock(GetIPCallbacks.class);
    getWiseFy().getIP(mockCallbacks);
    verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedIP(TEST_IP_ADDRESS_STRING);
  }
}

File: wisefy/src/androidTest/java/com/isupatches/wisefy/main/GetIIPTests.kt
package com.isupatches.wisefy.main

import com.isupatches.wisefy.BaseAndroidJUnit4TestClass
import com.isupatches.wisefy.TEST_IP_ADDRESS_STRING
import com.isupatches.wisefy.VERIFICATION_SUCCESS_TIMEOUT
import com.isupatches.wisefy.callbacks.GetIPCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER
import org.junit.Assert.assertEquals
import org.junit.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

/**
 * Tests the ability to get the IP of a device.
 *
 * @author Patches
 */
internal class GetIIPTests : BaseAndroidJUnit4TestClass() {

    @Test fun sync_getIP_failure_missingPrerequisites() {
        mockWiseFyPrechecksUtil.getIP_failure()
        assertEquals(null, wisefy.getIP())
    }

    @Test fun sync_getIP_failure() {
        mockWiseFyPrechecksUtil.getIP_success()
        mockNetworkUtil.ip_failure()
        assertEquals(null, wisefy.getIP())
    }

    @Test fun sync_getIP_success() {
        mockWiseFyPrechecksUtil.getIP_success()
        mockNetworkUtil.ip_success()
        assertEquals(TEST_IP_ADDRESS_STRING, wisefy.getIP())
    }

    @Test fun async_getIP_failure_missingPrerequisites() {
        mockWiseFyPrechecksUtil.getIP_failure()
        val mockCallbacks = mock(GetIPCallbacks::class.java)
        wisefy.getIP(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
    }

    @Test fun async_getIP_failure() {
        mockWiseFyPrechecksUtil.getIP_success()
        mockNetworkUtil.ip_failure()
        val mockCallbacks = mock(GetIPCallbacks::class.java)
        wisefy.getIP(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureRetrievingIP()
    }

    @Test fun async_getIP_success() {
        mockWiseFyPrechecksUtil.getIP_success()
        mockNetworkUtil.ip_success()
        val mockCallbacks = mock(GetIPCallbacks::class.java)
        wisefy.getIP(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedIP(TEST_IP_ADDRESS_STRING)
    }
}
